Improve NodeJS chapter (#182)
se-edu/learningresources: contents/node/nodejs.md

Author: Rachael Sim

<box type="info" light><md>

> This chapter assumes that the reader is familiar with JavaScript and asynchronous programming. If you are not familiar with asynchronous programming, a good resource to checkout is the [asynchronous programming section of the You Don't Know JS guide ](https://github.com/getify/You-Dont-Know-JS/blob/master/async%20&%20performance/README.md#you-dont-know-js-async--performance) as asynchronous programming is key in Node.</md>

Reviewer: Link appears to be broken

Reviewer 2: Link should be https://github.com/getify/You-Dont-Know-JS/tree/2nd-ed/sync-async ?

> ### Benefit 1: Easy to get Started

Reviewer 3: Easy to Get Started

Author: you didn't miss it, just merged the latest in 😃

> To install Node, simply download the installer from the [official Node website](https://nodejs.org/en/download/) based on your OS and run it. This should install both Node and npm. Npm is a tool which will help you to search, install and manage node packages. The use of npm will be explained [later](#benefit-5-easy-dependency-management-with-npm).

Reviewer: Instead of: "The use of npm will be explained [later]"

Consider: "We will explore npm further in the sections [below]"

> The following simple example demonstrates how a Node application imports required modules, creates a server to listen to client's request, and then reads the request and returns a response.

Reviewer 4: It sounds like this has 3 steps:
1. import stuff
1. create server
1. read request and return response

But step 3 obviously doesn't happen until a request is sent. Maybe try to rephrase?
```suggestion
The following simple example demonstrates how a Node application imports required modules, creates a server to listen for clients' requests, and then reads the request and returns a response.
```

> The following simple example demonstrates how a Node application imports required modules, creates a server to listen to client's request, and then reads the request and returns a response.

Reviewer: "The following simple example demonstrates..." sounds slightly awkward for me. I'd have written "The following is a simple example that demonstrates...". Small nitpick, up to you.

Also, I would change "listen to client's request..." to "listen to client requests followed by reading and returning a response for each".

> <box type="info" header="A Simple Hello World Server <small>from [codeburst](https://codeburst.io/getting-started-with-node-js-a-beginners-guide-b03e25bca71b)</small>" light>

Reviewer 4: The link is broken!

https://codeburst.io/getting-started-with-node-js-a-beginners-guide-b03e25bca71b

> <box type="info" header="A Simple Hello World Server <small>from [codeburst](https://codeburst.io/getting-started-with-node-js-a-beginners-guide-b03e25bca71b)</small>" light>

Reviewer: Link is broken as stated by @openorclose

```
To test the server, open a browser tab and navigate to http://localhost:3000/. You should see 'Hello world'.

</box>

Read on to find out about how to incorporate external dependencies, manage them and use Node's module system to organize your code.

### Benefit 2: Avoid Synchronization Problems and Overheads

<figure>
<img src="eventloop.png" width="74%" class="mx-auto mt-3 mb-2 d-block rounded" />

> <figcaption class="text-center"><small>

Reviewer: Consider mentioning or making reference to this figure in the following paragraph so it is not just "there" without anything talking about it.

Illustration of nodeJS's event loop from [these slides](https://www.slideshare.net/DevCentralAMD/introduction-to-nodejs-41469620)</small>
</figcaption>
</figure>

> Node is designed to be **event-driven**. When an <tooltip content="e.g. when a I/O operation is complete, or a timer fires">event</tooltip> occurs, the event handler previously registered with the event is queued to be run by the [event loop](https://medium.com/the-node-js-collection/what-you-should-know-to-really-understand-the-node-js-event-loop-and-its-metrics-c4907b19da4c). The main event loop (where Javascript code is executed) is single-threaded.

Reviewer 4: ```suggestion
Node is designed to be **event-driven**. When an <tooltip content="e.g. when a I/O operation is complete, or a timer fires">event</tooltip> occurs, the event handler previously registered with the event is queued to be run by the [event loop](https://medium.com/the-node-js-collection/what-you-should-know-to-really-understand-the-node-js-event-loop-and-its-metrics-c4907b19da4c). The main event loop (where JavaScript code is executed) is single-threaded.
```

> Node is designed to be **event-driven**. When an <tooltip content="e.g. when a I/O operation is complete, or a timer fires">event</tooltip> occurs, the event handler previously registered with the event is queued to be run by the [event loop](https://medium.com/the-node-js-collection/what-you-should-know-to-really-understand-the-node-js-event-loop-and-its-metrics-c4907b19da4c). The main event loop (where Javascript code is executed) is single-threaded.

Reviewer: Small nitpicks: 
- I think there is no need to bold "event-driven".
- Change "queued" to "enqueued"

> This means that we can avoid thread overheads and synchronization problems including deadlocks and race conditions.

Reviewer 4: ```suggestion
This means that we can avoid thread overheads and synchronization problems such as deadlocks and race conditions.
```

> This means that we can avoid thread overheads and synchronization problems including deadlocks and race conditions.

Reviewer: Consider adding links or a short explanation (similar to what you did for "event") for "deadlocks" and "race conditions".

> This means that we can avoid thread overheads and synchronization problems including deadlocks and race conditions.

Reviewer 2: Not sure if avoiding thread overheads by introducing event loop overheads is absolutely better, probably don't mention thread overheads?

Author: Depends on the use case; I think the context here was that most servers don't do computationally expensive work, where something like native C++ threads would really have an advantage

### Benefit 3: Fast for I/O Intensive Programs

I/O requests made in other languages such as Python or Java are typically blocking, which means the program remains idle (in the same thread) until the I/O operation completes.

> In contrast, nodeJS's allows for a **non-blocking** I/O model with its event-driven structure.

Reviewer 4: ```suggestion
In contrast, Node's allows for a **non-blocking** I/O model with its event-driven structure.
```

Reviewer: I think Node (without the apostrophe s) is fine too (and easier to read)

> I/O requests are delegated to other systems (e.g. file system and databases). While an I/O operation is incomplete, the event loop can still process subsequent requests. When the I/O request is complete, the handler registered with the request is scheduled to be executed on the event loop again.

Reviewer 4: ```suggestion
I/O requests are delegated to other systems (e.g. file system and databases). While an I/O operation is incomplete, the event loop can still process subsequent requests. When the I/O request is complete, the handler registered with the request is then scheduled to be executed on the event loop.
```

It's not registered again right? Only when when the operation has completed will the handler be scheduled for the first time.

### Benefit 4: Use JavaScript for Both Front and Back-end Development

> Using Node for back-end development makes it possible to share common code functionality between the front-end and back-end, which leads to less code maintainence. It also decreases the potential communication overhead between front and back-end developers, since they share the same code base, and may hence also lead to a more multidisciplinary team.

Reviewer: Consider replacing "It also decreases the potential..." with "It also reduces the potential..."

Consider appending the following after "... a more multidisciplinary team."
"that is familiar with both front and back-end development."

### Benefit 5: Easy Dependency Management With npm

Node Package Manager (npm) is used to

> * Search for node.js packages online

Reviewer 4: replace with Node

  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": ["promise", "lock"],
  "author": "",
  "license": "ISC"
}
```

> The **name** and a **version** forms a unique identifier for the package. When a package is updated, the version number must be updated. A good **description** string and **keywords** helps others in discovering your package.

Reviewer: Consider changing "helps others in discovering..." to "helps others discover..."

  }
}
```
The `dependencies` property maps to an object that has the name and
<popover header="**Version Range**" trigger="click">
<template slot="content">
It is common to find carets (`^`) and tildes (`~`) in the version range.<br>
This is part of npm's powerful <a href="https://nodesource.com/blog/semver-a-primer/">version specification system</a>, which allows

> users and developers of the package to keep its dependencies compatible.

Reviewer 2: Why newline here?

Author: just to keep the source neat 🙂

  }
}
```
Running `npm run build` in the command shell will execute `node app.js` and similarly `npm run lint-fix` will fix your linting errors in your JavaScript files.

Apart from acting as <tooltip content="See here for some common use cases of npm scripts!">[shortcuts](https://css-tricks.com/why-npm-scripts/)</tooltip> to commonly used commands, this also sets up and encourages a consistent development and build workflow in the project.

### Benefit 7: Easy to Reuse Code From Others

> Node module system allows you to include other Javascript files and thus makes it easy to reuse external libraries and organize your code into separate parts with limited responsibilities.

Reviewer 4: ```suggestion
Node's module system allows you to include other JavaScript files and thus makes it easy to reuse external libraries and organize your code into separate parts with limited responsibilities.
```

Node comes bundled with useful [core modules](https://nodejs.org/api/) such as the `fs` (file system) module which includes classes, methods and events to deal with file I/O operations and the `http` module which helps Node to transfer data over HTTP.

> There are also many useful and well-tested modules maintained by the community and external developers such as <tooltip content="A library providing various useful wrapper functions over javascript's promises">*Bluebird*</tooltip> on <tooltip content="the largest ecosystem of open source libraries in the world!">npm</tooltip>.

Reviewer 4: ```suggestion
There are also many useful and well-tested modules maintained by the community and external developers such as <tooltip content="A library providing various useful wrapper functions over JavaScript's promises">*Bluebird*</tooltip> on <tooltip content="the largest ecosystem of open source libraries in the world!">npm</tooltip>.
```

This makes development a breeze -- if a specific functionality has a large development overhead, you could look to such modules to speed things up.

> Importing modules is simple - simply use the `require()` function and provide the module identifier or the file path.

Reviewer: I would change "simple" to "easy" as it seems awkward to say "simple" followed by "simply".

```js
const http = require('http'); // import a core module
const Promise = require('bluebird'); // import a non-core module
```

> Node will first check if the module identifier passed to `require` is a core module or a relative path. If so, it will return the core module or the value of `module.exports` in the specified file path's code. Otherwise, Node will attempt to load the module from the `node_modules` folder in the parent directory of the current module.

Reviewer: I would phrase it the following way instead (changes in strike through and bold): 

Node will first check if the module identifier passed to `require` is a core module ~or a relative path~. If so, it will return the core module or the value of `module.exports` in the specified file path's code. Otherwise, Node will **treat the identifier as a relative path and** attempt to load the module from the `node_modules` folder in the parent directory of the current module.

Author: Hmm I think `relative path` here refers to file system relative paths, not module identifier relative paths, hence the original structure

### Benefit 8: Support for Better Code Organization

> With NodeJS's module system, you can create separate modules in our codebase such that each is focused on a single functionality. This makes your code more maintainable and testable.

Reviewer 4: ```suggestion
With NodeJS's module system, you can create separate modules in your codebase such that each is focused on a single functionality. This makes your code more maintainable and testable.
```

> For example, we could have a module for a parser. To export the `Parser` constructor in `parser.js`,

Reviewer 4: Change we to you? Seems like you is used in most places.

> For example, we could have a module for a parser. To export the `Parser` constructor in `parser.js`,

Reviewer 4: ```suggestion
For example, we could have a module for a parser. To export the `Parser` constructor in `parser.js`, you can assign `Parser` to `module.exports`.
```

Abrupt sentence end

```js
function Parser(options) {
  this._options = options || {};
}

Parser.prototype.parse = function (content) {
  ...
}
module.exports = Parser; // override the exports object
```

> Elsewhere, the Parser constructor would then be imported as such, allowing Parser objects to be created in other modules.

Reviewer: I would replace "as such," with "thus"

Author: `as such` here refers to the example right below it

Reviewer: Ah I see, my bad.

Author: No worries, thanks for the comprehensive review! 🙇

newParser.parse(content);
```

## Use Cases

Node is **good** for
* **Processing high volumes of I/O-bound requests**. A single instance of a Node server will be more efficient and can serve more requests with the same hardware than most other servers (such as those written in Ruby on Rails). This makes the Node server faster and more scalable.
* **Real time applications** where you have to process a large volume of requests with little delay. This includes instant messaging apps and collaborative editing apps where you can watch the document being modified live such as Trello and Google Docs. Node is a good choice as it can handle multiple client requests even while waiting for responses.

> * **Single-page applications** where a lot of processing and rendering is done on the client's side and the backend server only needs to provide a simple API. Node can process many requests with low response times. In addition, you can reuse and share code, such as those for validation, between the client and server.

Reviewer 2: ```suggestion
* **Single-page applications** where a lot of processing and rendering is done on the client's side and the backend server only needs to provide a simple API. Node can process many requests with low response times. In addition, you can reuse and share code written in JavaScript, such as those for validation, between the client and server.
```

However, Node is **not suitable** for

> * **CPU-intensive jobs**. Recall that Node is single-threaded. If the thread is busy doing CPU-heavy operations, it will not be able to process incoming requests timely. This removes the biggest advantage of Node. In such cases, you could consider alternatives such as Go which supports concurrency.

Reviewer 2: "concurrency" technically means executing processes in a (pseudo-)parallel manner and Node supports concurrency by nature. I guess the author meant to say that Node is after all a JavaScript runtime that's not as performant as a native compiled language like Go, and CPU-intensive tasks won't execute as efficiently in Node as Go.

Author: Added the idea of native threads here, thanks!
I think worker threads weren't out when this article was written 🤣

* Try out Node online - [Node prototyping with Runkit](https://runkit.com/home)

> * A guide on asynchronous programming in JS - [You Don't Know JS: Async & Performance](https://github.com/getify/You-Dont-Know-JS/blob/master/async%20&%20performance/README.md#you-dont-know-js-async--performance)

Reviewer 4: dead link

* A comprehensive introduction to Node -  [The art of node](https://github.com/maxogden/art-of-node/)
* Making better use of npm scripts - [freeCodeCamp's Introduction to NPM Scripts](https://medium.freecodecamp.org/introduction-to-npm-scripts-1dbb2ae01633), [How to Use npm as a Build Tool](https://www.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/)
* Using `module.exports` - [Tendai Mutunhire's article](http://stackabuse.com/how-to-use-module-exports-in-node-js/).
* Advice on how to write clean code that makes it easy to add new features - [Fundamental rules of a Node.js project structure](https://blog.risingstack.com/node-hero-node-js-project-structure-tutorial/)
* A compilation of useful node modules - [Awesome Nodejs](https://github.com/sindresorhus/awesome-nodejs)

> *  A summary and curation of the top-ranked content on Node best practices - [Node.js Best Practices](https://github.com/i0natan/nodebestpractices)

Reviewer 4: * Node architecture and features - [All About Node.Js You Wanted To Know](https://codeburst.io/all-about-node-js-you-wanted-to-know-25f3374e0be7)

dead link
